internal/datastore/crdb/pool/balancer.go

```diff
@@ -0,0 +1,214 @@
+)
+
+var connectionsPerCRDBNodeCountGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	Name: "crdb_connections_per_node",
+	Help: "the number of connections spicedb has to each crdb node",
+}, []string{"pool", "node_id"})
+
+func init() {
+	prometheus.MustRegister(connectionsPerCRDBNodeCountGauge)
+}
+
+type balancePoolConn[C balanceConn] interface {
+	Conn() C
+	Release()
+}
+
+type balanceConn interface {
+	comparable
+	IsClosed() bool
+}
+
+// balanceablePool is an interface that a pool must implement to allow its
+// connections to be balanced by the balancer.
+type balanceablePool[P balancePoolConn[C], C balanceConn] interface {
+	ID() string
+	AcquireAllIdle(ctx context.Context) []P
+	Node(conn C) uint32
+	GC(conn C) uint32
+	MaxConns() uint32
+	Range(func(conn C, nodeID uint32))
+}
+
+// NodeConnectionBalancer attempts to keep the connections managed by a RetryPool balanced between healthy nodes in
+// a Cockroach cluster.
+// It asynchronously processes idle connections, and kills any to nodes that have too many. When the pool reconnects,
+// it will have a different balance of connections, and over time the balancer will bring the counts close to equal.
+type NodeConnectionBalancer struct {
+	nodeConnectionBalancer[*pgxpool.Conn, *pgx.Conn]
+}
+
+// NewNodeConnectionBalancer builds a new nodeConnectionBalancer for a given connection pool and health tracker.
+func NewNodeConnectionBalancer(pool *RetryPool, healthTracker *NodeHealthTracker, interval time.Duration) *NodeConnectionBalancer {
+	return &NodeConnectionBalancer{*newNodeConnectionBalancer[*pgxpool.Conn, *pgx.Conn](pool, healthTracker, interval)}
+}
+
+// nodeConnectionBalancer is generic over underlying connection types for
+// testing purposes. Callers should use the exported NodeConnectionBalancer
+type nodeConnectionBalancer[P balancePoolConn[C], C balanceConn] struct {
+	ticker        *time.Ticker
+	pool          balanceablePool[P, C]
+	healthTracker *NodeHealthTracker
+	rnd           *rand.Rand
+	seed          int64
+}
+
+// newNodeConnectionBalancer is generic over underlying connection types for
+// testing purposes. Callers should use the exported NewNodeConnectionBalancer.
+func newNodeConnectionBalancer[P balancePoolConn[C], C balanceConn](pool balanceablePool[P, C], healthTracker *NodeHealthTracker, interval time.Duration) *nodeConnectionBalancer[P, C] {
+	seed := int64(new(maphash.Hash).Sum64())
+	return &nodeConnectionBalancer[P, C]{
+		ticker:        time.NewTicker(interval),
+		healthTracker: healthTracker,
+		pool:          pool,
+		seed:          seed,
+		rnd:           rand.New(rand.NewSource(seed)),
+	}
+}
+
+// Prune starts periodically checking idle connections and killing ones that are determined to be unbalanced.
+func (p *nodeConnectionBalancer[P, C]) Prune(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			p.ticker.Stop()
+			return
+		case <-p.ticker.C:
+			p.pruneConnections(ctx)
+			runtime.GC()
```

Comment:
Removed! I had some versions of this that attached a finalizer to the pgx.Conn because it seemed like the pool was removing connections without calling `BeforeClose`. but as mentioned out of band, the actual issue was the alternate map implementation I had been using (that's no longer in this PR at all), and there's no need for the finalizer. This is leftover from that; removed it.